README.md

```diff
@@ -0,0 +1,57 @@
+# @xmtp/snap
+
+This repository contains the XMTP Snap, released via NPM here, as well as a small webapp to demo the functionality. It was generated using the [template-snap repository](https://github.com/MetaMask/template-snap-monorepo/generate)
+
+## Architecture
+
+The XMTP Snap is an implementation of the Keystore API, a defined interface for XMTP clients to interact with a Keystore holding XMTP key material.
+
+### RPC Code Generation
+
+The Keystore API is defined in Protobuf, but Snaps are required to communicate via JSON-RPC. To allow clients to communicate with the Snap Keystore, all request protos are serialized and base64 encoded. In the snap, requests are base64 decoded and deserialized and then responses are once again serialized and base64 encoded.
+
+This is done automatically, using [RPC definitions defined in `xmtp-js`](https://github.com/xmtp/xmtp-js/blob/snap/src/keystore/rpcDefinitions.ts) that define expected request and response types for each method.
+
+### Authentication
+
+Two methods in the Snap allow unauthenticated access. `initKeystore` and `getKeystoreStatus`.
+
+`initKeystore` takes an XMTP `PrivateKeyBundle` as an argument and saves it in the Snaps storage. Upon successful validation and storage of the XMTP keys, the origin that called `initKeystore` is authorized to make calls to restricted Keystore methods for 30 days.
+
+`getKeystoreStatus` allows the caller to check if keys are present in the Snap storage for a given wallet address/environment combination. Clients are expected to call this method at the beginning of a session to see if they need to call generate/load the keys themselves and call `initKeystore` or simply proceed using the already stored keys.
```

Comment:
Wondering if `getKeystoreStatus` could be abused to collect data on wallet usage on the network. Couldn't it be also gated by the same authentication requirements as the rest of the API. An app can still probe whether it already created the key, if it gets an auth error it didn't, no?

Reply:
`getKeystoreStatus` wouldn't be able to collect data on wallet usage across the network. It can only tell you if the currently installed Metamask Snap has a key for a given wallet address/env. To access the API, the user has to have gone through the installation process for the Snap first on that domain (so it can't be checked in the background in every app without the user noticing).

It seems like checking for auth errors would get you the same result as using the API unauthenticated, wouldn't it?

Reply:
I probably miss important parts of the picture, but I thought that given that it's an RPC based access that unauthorized sites could call the RPC endpoint. If that's not possible then I agree that there's nothing to worry about.

With the auth errors, I assumed that unauthorized app would get an auth error whether the wallet/key is registered or not, so it shouldn't be able to learn anything.

Reply:
> but I thought that given that it's an RPC based access that unauthorized sites could call the RPC endpoint. If that's not possible then I agree that there's nothing to worry about.

We should be protected from that, as a given origin has to have the snap installed and approved before it can do anything. Our authorization system is a second layer on top of that granting permissions at the per wallet/env level.

I'll update the README to clarify